File: 1657.determine-if-two-strings-are-close.py
#
# @lc app=leetcode id=1657 lang=python3
#
# [1657] Determine if Two Strings Are Close
#

# @lc code=start
class Solution:
    def closeStrings(self, word1: str, word2: str) -> bool:
        map1 , map2 = {}, {}
        char_set1, char_set2 = set(), set()
        for c in word1:
            map1[c] = map1.get(c, 0) + 1
            char_set1.add(c)
        for c in word2:
            map2[c] = map2.get(c, 0) + 1
            char_set2.add(c)
        
        for c in char_set1:
            if c not in char_set2:
                return False
        
        count_count1, count_count2 = {}, {}
        
        for value in map1.values():
            count_count1[value] = count_count1.get(value, 0) + 1
        
        for value in map2.values():
            count_count2[value] = count_count2.get(value, 0) + 1
        
        for key, value in count_count1.items():
            if key not in count_count2:
                return False
            if count_count2[key] != value:
                return False
        
        return True
        
# @lc code=end

File: 1657.determine-if-two-strings-are-close.cpp
/*
 * @lc app=leetcode id=1657 lang=cpp
 *
 * [1657] Determine if Two Strings Are Close
 */
#include <bits/stdc++.h>
using namespace std;
// @lc code=start
class Solution
{
public:
    bool closeStrings(string word1, string word2)
    {
        vector<int> freq1(26, 0), freq2(26, 0);

        for (const char &c : word1)
        {
            freq1[c - 'a']++;
        }
        for (const char &c : word2)
        {
            freq2[c - 'a']++;
        }

        for (int i = 0; i < 26; ++i)
        {
            bool has1, has2;
            has1 = (bool)freq1[i];
            has2 = (bool)freq2[i];

            if (has1 ^ has2)
            {
                return false;
            }
        }

        sort(begin(freq1), end(freq1));
        sort(begin(freq2), end(freq2));

        return freq1 == freq2;
    }
};
// @lc code=end
